File: src/com/premiere/Viewer.java
package com.premiere;

/**
 * Клас Viewer реалізує роль глядача у системі.
 * Демонструє наслідування від User та реалізацію Engageable інтерфейсу.
 */
public class Viewer extends User implements Engageable {
    // Приватні атрибути - інкапсуляція
    private boolean subscribed;
    private int watchedCount;

    /**
     * Конструктор для ініціалізації глядача
     */
    public Viewer(String name, String email) {
        super(name, email);
        this.subscribed = false;
        this.watchedCount = 0;
    }

    /**
     * Реалізація методу для підписки на прем'єру
     */
    @Override
    public void subscribe() {
        this.subscribed = true;
        notifySystem("Підписався на прем'єру");
    }

    /**
     * Реалізація методу для перегляду контенту
     */
    @Override
    public void watchContent(Content content) {
        if (subscribed) {
            content.addView();
            this.watchedCount++;
            notifySystem("Переглядає відео: \"" + content.getTitle() + "\"");
        } else {
            notifySystem("Не може переглядати, не підписаний на прем'єру");
        }
    }

    /**
     * Реалізація методу для коментування контенту
     */
    @Override
    public void commentOnContent(String comment) {
        notifySystem("Залишив коментар: \"" + comment + "\"");
    }

    /**
     * Реалізація абстрактного методу performAction
     */
    @Override
    public void performAction() {
        System.out.println("[ДІЯ ГЛЯДАЧА]: " + name + " взаємодіє з контентом (переглядає, коментує)");
    }

    /**
     * Геттер для перевірки статусу підписки
     */
    public boolean isSubscribed() {
        return subscribed;
    }

    /**
     * Геттер для отримання кількості переглянутих відео
     */
    public int getWatchedCount() {
        return watchedCount;
    }
}

File: src/Premiere/Viewer.cs
using System;

namespace Premiere
{
    /// <summary>
    /// Клас Viewer реалізує роль глядача у системі.
    /// Демонструє наслідування від User та реалізацію IEngageable інтерфейсу.
    /// </summary>
    public class Viewer : User, IEngageable
    {
        // Приватні атрибути - інкапсуляція
        private bool Subscribed { get; set; }
        private int WatchedCount { get; set; }

        /// <summary>
        /// Конструктор для ініціалізації глядача
        /// </summary>
        public Viewer(string name, string email) : base(name, email)
        {
            Subscribed = false;
            WatchedCount = 0;
        }

        /// <summary>
        /// Реалізація методу для підписки на прем'єру
        /// </summary>
        public void Subscribe()
        {
            Subscribed = true;
            NotifySystem("Підписався на прем'єру");
        }

        /// <summary>
        /// Реалізація методу для перегляду контенту
        /// </summary>
        public void WatchContent(Content content)
        {
            if (Subscribed)
            {
                content.AddView();
                WatchedCount++;
                NotifySystem($"Переглядає відео: \"{content.GetTitle()}\"");
            }
            else
            {
                NotifySystem("Не може переглядати, не підписаний на прем'єру");
            }
        }

        /// <summary>
        /// Реалізація методу для коментування контенту
        /// </summary>
        public void CommentOnContent(string comment)
        {
            NotifySystem($"Залишив коментар: \"{comment}\"");
        }

        /// <summary>
        /// Реалізація абстрактного методу PerformAction
        /// </summary>
        public override void PerformAction()
        {
            Console.WriteLine($"[ДІЯ ГЛЯДАЧА]: {Name} взаємодіє з контентом (переглядає, коментує)");
        }

        /// <summary>
        /// Метод для перевірки статусу підписки
        /// </summary>
        public bool IsSubscribed()
        {
            return Subscribed;
        }

        /// <summary>
        /// Метод для отримання кількості переглянутих відео
        /// </summary>
        public int GetWatchedCount()
        {
            return WatchedCount;
        }
    }
}
